File: sportsbook-app/payment-validator-service/src/main/java/com/sportsbook/payment/validator/service/ValidationOrchestrator.java
package com.sportsbook.payment.validator.service;

import com.sportsbook.payment.contracts.*;
import com.sportsbook.payment.contracts.saga.*;
import com.sportsbook.payment.validator.domain.DepositDecisionEntity;
import com.sportsbook.payment.validator.domain.DepositDecisionRepository;
import com.sportsbook.payment.validator.ports.CreditCheckPort;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.Optional;
import java.util.UUID;

@Service
public class ValidationOrchestrator {

    private final CreditCheckPort creditCheckPort;
    private final DepositDecisionRepository decisionRepository;
    private final KafkaTemplate<String, Object> eventsKafkaTemplate;

    private final String validatedTopic = "deposit-validated";
    private final String rejectedTopic = "deposit-rejected";
    private final String commandsTopic = "deposit-commands";
    private final String eventsTopic = "deposit-events";

    public ValidationOrchestrator(CreditCheckPort creditCheckPort,
                                  DepositDecisionRepository decisionRepository,
                                  KafkaTemplate<String, Object> eventsKafkaTemplate) {
        this.creditCheckPort = creditCheckPort;
        this.decisionRepository = decisionRepository;
        this.eventsKafkaTemplate = eventsKafkaTemplate;
    }

    public void validateAndOrchestrate(DepositEnrichedEvent enrichedEvent) {

        var deposit = enrichedEvent.original();

        Optional<DepositDecisionEntity> existing = decisionRepository.findById(deposit.depositId());
        if (existing.isPresent()) {
            return;
        }

        FundingSourceType sourceType = creditCheckPort.checkSource(enrichedEvent);

        DepositDecision decision;
        DepositRejectionReason reason = null;

        switch (sourceType) {
            case DEBIT -> decision = DepositDecision.APPROVED;
            case CREDIT -> {
                decision = DepositDecision.REJECTED;
                reason = DepositRejectionReason.CREDIT_CARD;
            }
            case OVERDRAFT -> {
                decision = DepositDecision.REJECTED;
                reason = DepositRejectionReason.OVERDRAFT;
            }
            case BNPL -> {
                decision = DepositDecision.REJECTED;
                reason = DepositRejectionReason.BNPL_PROVIDER;
            }
            default -> {
                decision = DepositDecision.REJECTED;
                reason = DepositRejectionReason.AMBIGUOUS_BIN;
            }
        }

        var validatedEvent = new DepositValidatedEvent(
                deposit.depositId(),
                deposit.playerId(),
                deposit.amount(),
                sourceType,
                decision,
                reason,
                null,
                Instant.now()
        );

        var entity = new DepositDecisionEntity();
        entity.setDepositId(validatedEvent.depositId());
        entity.setPlayerId(validatedEvent.playerId());
        entity.setAmount(validatedEvent.amount());
        entity.setFundingSourceType(validatedEvent.fundingSourceType());
        entity.setDecision(validatedEvent.decision());
        entity.setRejectionReason(validatedEvent.rejectionReason());
        entity.setProviderReference(validatedEvent.providerReference());
        entity.setDecidedAt(validatedEvent.decidedAt());
        decisionRepository.save(entity);

        if (decision == DepositDecision.APPROVED) {
            eventsKafkaTemplate.send(validatedTopic, deposit.playerId(), validatedEvent);

            String sagaId = UUID.randomUUID().toString();
            var sagaEvent = new DepositValidatedSagaEvent(sagaId, validatedEvent);
            eventsKafkaTemplate.send(eventsTopic, validatedEvent.depositId(), sagaEvent);

            var creditWalletCommand = new CreditWalletCommand(
                    sagaId,
                    validatedEvent.depositId(),
                    validatedEvent.playerId(),
                    validatedEvent.amount()
            );
            eventsKafkaTemplate.send(commandsTopic, validatedEvent.playerId(), creditWalletCommand);

        } else {
            eventsKafkaTemplate.send(rejectedTopic, deposit.playerId(), validatedEvent);

            String sagaId = UUID.randomUUID().toString();
            var sagaEvent = new DepositValidatedSagaEvent(sagaId, validatedEvent);
            eventsKafkaTemplate.send(eventsTopic, validatedEvent.depositId(), sagaEvent);

            var releaseCommand = new ReleaseReservationCommand(
                    sagaId,
                    validatedEvent.depositId(),
                    "psp-reservation-id-placeholder"
            );
            eventsKafkaTemplate.send(commandsTopic, validatedEvent.depositId(), releaseCommand);
        }
    }
}

File: sportsbook-app/payment-validator-service/src/main/kotlin/com/sportsbook/payment/validator/service/ValidationOrchestrator.kt
package com.sportsbook.payment.validator.service

import com.sportsbook.payment.contracts.DepositDecision
import com.sportsbook.payment.contracts.DepositEnrichedEvent
import com.sportsbook.payment.contracts.DepositRejectionReason
import com.sportsbook.payment.contracts.DepositValidatedEvent
import com.sportsbook.payment.contracts.FundingSourceType
import com.sportsbook.payment.contracts.saga.CreditWalletCommand
import com.sportsbook.payment.contracts.saga.DepositValidatedSagaEvent
import com.sportsbook.payment.contracts.saga.ReleaseReservationCommand
import com.sportsbook.payment.validator.domain.DepositDecisionEntity
import com.sportsbook.payment.validator.domain.DepositDecisionRepository
import com.sportsbook.payment.validator.ports.CreditCheckPort
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service
import java.time.Instant
import java.util.UUID

@Service
class ValidationOrchestrator(
    private val creditCheckPort: CreditCheckPort,
    private val decisionRepository: DepositDecisionRepository,
    private val eventsKafkaTemplate: KafkaTemplate<String, Any>,
) {

    private val validatedTopic = "deposit-validated"
    private val rejectedTopic = "deposit-rejected"
    private val commandsTopic = "deposit-commands"
    private val eventsTopic = "deposit-events"

    fun validateAndOrchestrate(enrichedEvent: DepositEnrichedEvent) {
        val deposit = enrichedEvent.original

        if (decisionRepository.findById(deposit.depositId).isPresent) {
            return
        }

        val sourceType = creditCheckPort.checkSource(enrichedEvent)

        val decision: DepositDecision
        var reason: DepositRejectionReason? = null

        when (sourceType) {
            FundingSourceType.DEBIT -> decision = DepositDecision.APPROVED
            FundingSourceType.CREDIT -> {
                decision = DepositDecision.REJECTED
                reason = DepositRejectionReason.CREDIT_CARD
            }
            FundingSourceType.OVERDRAFT -> {
                decision = DepositDecision.REJECTED
                reason = DepositRejectionReason.OVERDRAFT
            }
            FundingSourceType.BNPL -> {
                decision = DepositDecision.REJECTED
                reason = DepositRejectionReason.BNPL_PROVIDER
            }
            else -> {
                decision = DepositDecision.REJECTED
                reason = DepositRejectionReason.AMBIGUOUS_BIN
            }
        }

        val validatedEvent = DepositValidatedEvent(
            deposit.depositId,
            deposit.playerId,
            deposit.amount,
            sourceType,
            decision,
            reason,
            null,
            Instant.now(),
        )

        val entity = DepositDecisionEntity().apply {
            depositId = validatedEvent.depositId
            playerId = validatedEvent.playerId
            amount = validatedEvent.amount
            fundingSourceType = validatedEvent.fundingSourceType
            this.decision = validatedEvent.decision
            rejectionReason = validatedEvent.rejectionReason
            providerReference = validatedEvent.providerReference
            decidedAt = validatedEvent.decidedAt
        }
        decisionRepository.save(entity)

        if (decision == DepositDecision.APPROVED) {
            eventsKafkaTemplate.send(validatedTopic, deposit.playerId, validatedEvent)

            val sagaId = UUID.randomUUID().toString()
            val sagaEvent = DepositValidatedSagaEvent(sagaId, validatedEvent)
            eventsKafkaTemplate.send(eventsTopic, validatedEvent.depositId, sagaEvent)

            val creditWalletCommand = CreditWalletCommand(
                sagaId,
                validatedEvent.depositId,
                validatedEvent.playerId,
                validatedEvent.amount,
            )
            eventsKafkaTemplate.send(commandsTopic, validatedEvent.playerId, creditWalletCommand)
        } else {
            eventsKafkaTemplate.send(rejectedTopic, deposit.playerId, validatedEvent)

            val sagaId = UUID.randomUUID().toString()
            val sagaEvent = DepositValidatedSagaEvent(sagaId, validatedEvent)
            eventsKafkaTemplate.send(eventsTopic, validatedEvent.depositId, sagaEvent)

            val releaseCommand = ReleaseReservationCommand(
                sagaId,
                validatedEvent.depositId,
                "psp-reservation-id-placeholder",
            )
            eventsKafkaTemplate.send(commandsTopic, validatedEvent.depositId, releaseCommand)
        }
    }
}
